File: languages/swift/reverse-complement.swift
// reverse-complement: generate a DNA sequence, reverse it in place while complementing
// each base (A<->T, C<->G), then reduce it to a polynomial string hash. The reverse uses a
// hand-written two-pointer loop (NOT a stdlib bulk reverse) and the hash a per-character
// loop (NOT a builtin), so this measures the language's own per-character processing -
// consistent with the suite's no-stdlib-shortcut rule. Everything is integer-deterministic.
import Foundation

let P: Int = 1000000007
let IM: Int = 139968
let IA: Int = 3877
let IC: Int = 29573

// ASCII byte values: A=65, C=67, G=71, T=84
let A: UInt8 = 65
let C: UInt8 = 67
let G: UInt8 = 71
let T: UInt8 = 84

func comp(_ c: UInt8) -> UInt8 {    // A<->T, C<->G; only A/C/G/T occur
    if c == A { return T }
    if c == C { return G }
    if c == G { return C }
    return A
}

func run(_ L: Int) -> Int {
    var s = [UInt8](repeating: 0, count: L)
    var seed = 42
    for i in 0..<L {
        seed = (seed * IA + IC) % IM
        s[i] = seed < 42000 ? A : seed < 70000 ? C : seed < 98000 ? G : T
    }
    var i = 0, j = L - 1
    while i < j {                    // two-pointer reverse-and-complement, in place
        let a = comp(s[i])
        s[i] = comp(s[j])
        s[j] = a
        i += 1; j -= 1
    }
    if i == j { s[i] = comp(s[i]) }  // middle char when L is odd
    var h = 0
    for k in 0..<L {
        h = (h * 31 + Int(s[k])) % P
    }
    return h
}

let L = CommandLine.arguments.count > 1 ? (Int(CommandLine.arguments[1]) ?? 100000) : 100000
print(run(L))
print("reverse-complement(\(L))")

File: languages/scala/reverse-complement.scala
object ReverseComplement {
  final val P  = 1000000007L
  final val IM = 139968L
  final val IA = 3877L
  final val IC = 29573L

  def comp(c: Byte): Byte =            // A<->T, C<->G; only A/C/G/T occur
    if (c == 'A') 'T' else if (c == 'C') 'G' else if (c == 'G') 'C' else 'A'

  def reverseComplement(L: Int): Long = {
    val s = new Array[Byte](L)
    var seed = 42L
    var i = 0
    while (i < L) {
      seed = (seed * IA + IC) % IM
      s(i) =
        if (seed < 42000) 'A'
        else if (seed < 70000) 'C'
        else if (seed < 98000) 'G'
        else 'T'
      i += 1
    }

    i = 0
    var j = L - 1
    while (i < j) {                    // two-pointer reverse-and-complement, in place
      val a = comp(s(i))
      s(i) = comp(s(j))
      s(j) = a
      i += 1; j -= 1
    }
    if (i == j) s(i) = comp(s(i))      // middle char when L is odd

    var h = 0L
    var k = 0
    while (k < L) {
      h = (h * 31 + (s(k) & 0xff)) % P // ASCII byte value: A=65, C=67, G=71, T=84
      k += 1
    }
    h
  }

  def main(args: Array[String]): Unit = {
    val L = if (args.nonEmpty) args(0).toInt else 100000
    println(reverseComplement(L))
    println(s"reverse-complement($L)")
  }
}
